python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,1010 @@
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    monkeypatch.setattr(ray.dag.constants, "RAY_ADAG_ENABLE_DETECT_DEADLOCK", False)
+
+    a = Worker.remote()
+    b = Worker.remote()
+    c = Worker.remote()
+
+    with InputNode() as inp:
+        branch1 = a.no_op.bind(inp)
+        branch1.with_type_hint(TorchTensorType(transport="nccl"))
+        branch2 = b.no_op.bind(inp)
+        branch2.with_type_hint(TorchTensorType(transport="nccl"))
+        branch3 = c.no_op.bind(inp)
+        branch3.with_type_hint(TorchTensorType(transport="nccl"))
+        dag = MultiOutputNode(
+            [
+                a.no_op.bind(branch3),
+                b.no_op.bind(branch1),
+                c.no_op.bind(branch2),
+            ]
+        )
+
+    compiled_dag = dag.experimental_compile()
+
+    a_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+    b_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+    c_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.READ),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+
+    a_schedule = compiled_dag.actor_to_execution_schedule[a]
+    b_schedule = compiled_dag.actor_to_execution_schedule[b]
+    c_schedule = compiled_dag.actor_to_execution_schedule[c]
+
+    for schedule, expected_schedule in zip(
+        [a_schedule, b_schedule, c_schedule],
+        [a_expected_schedule, b_expected_schedule, c_expected_schedule],
+    ):
+        assert len(schedule) == len(expected_schedule)
+        for i, operation in enumerate(schedule):
+            assert operation.idx == expected_schedule[i][0]
+            assert operation.type == expected_schedule[i][1]
+
+    tensor_cpu = torch.zeros(10, 10)
+    ref = compiled_dag.execute(tensor_cpu)
+    tensors = ray.get(ref)
+    tensor_cuda = tensor_cpu.to("cuda:0")
+
+    assert len(tensors) == 3
+    for t in tensors:
+        assert torch.equal(t, tensor_cuda)
+
+    compiled_dag.teardown()
+
+
+def generate_dag_graph_nodes(local_idx, global_idx, actor_handle, requires_nccl):
+    graph_nodes = {}
+    for op_type in DAGNodeOperationType:
+        graph_nodes[op_type] = DAGOperationGraphNode(
+            DAGNodeOperation(local_idx, op_type),
+            global_idx,
+            actor_handle,
+            requires_nccl,
+        )
+    return graph_nodes
+
+
+class TestSelectNextNodes:
```

Comment:
for these unit tests, can we run it in non-gpu tests?

Reply:
Yes, let's move all unit tests (i.e. tests that don't actually create actors) to a separate test suite.

Also, can you see if you can add unit tests that check the final produced schedule, but without having to create the compiled DAG?

Please also add unit tests to check that for DAGs without any NCCL operations, we simply follow the bind index in the final schedule.

Reply:
Split tests into CPU / GPU tests: https://github.com/ray-project/ray/pull/46911/commits/72799f791151bb3774a424f11cdefad63d239026

Reply:
> Also, can you see if you can add unit tests that check the final produced schedule, but without having to create the compiled DAG? 

@stephanie-wang Is `TestGenerateActorToExecutionSchedule` what you are referring to?